### PR TITLE
Improved API support for Firebase Database.

### PR DESCRIPTION
**Pull request by collinjackson**
Fixes flutter/flutter#10411 
Fixes flutter/flutter#10421

### CODE REVIEW

**Comment by eseidelGoogle** [comment]
Some of @lukef and @KevinTheGray's firebase work is believed to be blocked on this. :)  They're considering pushing off firebase work for a week if this needs more time to bake.  Just checking in.

**Comment by collinjackson** [comment]
Most of the database code I've been working on has been Database UI improvements (fixes flutter/flutter#10439) and an improved example app. I'd like another day or so to test those more prior to landing.

Based on flutter/flutter#10411 and flutter/flutter#10421 it sounds like @lukef and@KevinTheGray mostly just need the core database additions. So I'm punting all the UI changes into a separate PR and this PR is now just about adding the missing APIs (including tests).

I think this can land as soon as I get an LGTM on it.

File: packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java [hunk 1]
```diff
@@ -118,38 +118,101 @@ public void onDataChange(DataSnapshot snapshot) {
...
-      EventObserver observer = observers.get(handle);
-      if (observer != null) {
-        if (observer.requestedEventType.equals(EVENT_TYPE_VALUE)) {
-          reference.removeEventListener((ValueEventListener) observer);
-        } else {
-          reference.removeEventListener((ChildEventListener) observer);
+    Map<String, Object> arguments = (Map<String, Object>) call.arguments;
+    switch (call.method) {
+      case "FirebaseDatabase#goOnline":
+        {
```
Comment (goderbauer) [inline packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java]: Does this need to be in a block?

Comment (collinjackson) [inline packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java]: It doesn't, I was trying to be consistent with the others but maybe it's better not to do a block. I've removed the braces for the methods that don't take arguments.

File: packages/firebase_database/lib/src/database_reference.dart [hunk 2]
```diff
@@ -72,12 +72,53 @@ class DatabaseReference extends Query {
...
+  /// Children are sorted based on this priority using the following rules:
+  ///
+  /// Children with no priority come first. Children with a number as their
+  /// priority come next. They are sorted numerically by priority (small to
+  /// large). Children with a string as their priority come last. They are
+  /// sorted lexicographically by priority. Whenever two children have the same
+  /// priority (including no priority), they are sorted by key. Numeric keys
+  /// come first (sorted numerically), followed by the remaining keys (sorted
+  /// lexicographically).
+
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/database_reference.dart]: Missing `///` here?

File: packages/firebase_database/lib/src/database_reference.dart [hunk 3]
```diff
@@ -72,12 +72,53 @@ class DatabaseReference extends Query {
...
+  /// Note that priorities are parsed and ordered as IEEE 754 double-precision
+  /// floating-point numbers. Keys are always stored as strings and are treated
+  /// as numbers only when they can be parsed as a 32-bit integer
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/database_reference.dart]: [nit] Missing `.` at the end.

File: packages/firebase_database/lib/src/firebase_database.dart [hunk 4]
```diff
@@ -22,17 +23,73 @@ class FirebaseDatabase {
...
   /// Gets a DatabaseReference for the root of your Firebase Database.
   DatabaseReference reference() => new DatabaseReference._(this, <String>[]);
+
+  /// The Firebase Database client will cache synchronized data and keep track
+  /// of all writes you’ve initiated while your application is running. It
+  /// seamlessly handles intermittent network connections and re-sends write
+  /// operations when the network connection is restored.
+  ///
+  /// However by default your write operations and cached data are only stored
+  /// in-memory and will be lost when your app restarts. By setting this value
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/firebase_database.dart]: `this value to YES` => `[enabled] to true`

File: packages/firebase_database/lib/src/firebase_database.dart [hunk 5]
```diff
@@ -22,17 +23,73 @@ class FirebaseDatabase {
...
+  /// to YES, the data will be persisted to on-device (disk) storage and will
+  /// thus be available again when the app is restarted (even when there is no
+  /// network connectivity at that time). Note that this property must be set
+  /// before creating your first Database reference and only needs to be called
+  /// once per application.
+  Future<Null> setPersistenceEnabled(bool enabled) {
+    return _channel.invokeMethod(
+      "FirebaseDatabase#setPersistenceEnabled",
+      { 'enabled': enabled },
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/firebase_database.dart]: Why not just send the boolean `enabled` as argument? Do we expect more parameters in a later revision?

Similarly for other methods.

File: packages/firebase_database/lib/src/firebase_database.dart [hunk 6]
```diff
@@ -22,17 +23,73 @@ class FirebaseDatabase {
...
+  /// or greater than 100 MB are not supported.
+  Future<Null> setPersistenceCacheSizeBytes(int cacheSize) {
+    return _channel.invokeMethod(
+      "FirebaseDatabase#setPersistenceCacheSizeBytes",
+      { 'cacheSize': cacheSize },
+    );
+  }
+
+  /// Resumes our connection to the Firebase Database backend after a previous
+  /// goOffline call.
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/firebase_database.dart]: `[goOffline]`

File: packages/firebase_database/lib/src/firebase_database.dart [hunk 7]
```diff
@@ -22,17 +23,73 @@ class FirebaseDatabase {
...
+  Future<Null> goOnline() {
+    return _channel.invokeMethod("FirebaseDatabase#goOnline");
+  }
+
+  /// Shuts down our connection to the Firebase Database backend until goOnline
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/firebase_database.dart]: `[goOnline]`

File: packages/firebase_database/lib/src/query.dart [hunk 8]
```diff
@@ -58,4 +86,90 @@ class Query {
...
+    return _copyWithParameters({ 'startAt': value, 'startAtKey': key});
+  }
+
+  /// Create a query constrained to only return child nodes with a value less
+  /// than or equal to the given value, using the given orderBy directive or
+  /// priority as default, and optionally only child nodes with a key less
+  /// than or equal to the given key.
+  Query endAt(dynamic value, { String key }) {
+    assert(!_parameters.containsKey('endAt'));
+    return _copyWithParameters({ 'endAt': value, 'endAtKey': key});
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/query.dart]: [nit] Missing space before `}`

File: packages/firebase_database/lib/src/query.dart [hunk 9]
```diff
@@ -58,4 +86,90 @@ class Query {
...
+
+  /// Create a query with limit and anchor it to the end of the window.
+  Query limitToLast(int limit) {
+    assert(!_parameters.containsKey('limitToLast'));
+    return _copyWithParameters({ 'limitToLast': limit });
+  }
+
+  /// Generate a view of the data sorted by values of a particular child key.
+  ///
+  /// Intended to be used in combination with startAt(), endAt(), or equalTo().
```
Comment (mravn-google) [inline packages/firebase_database/lib/src/query.dart]: `[startAt()]`, `[endAt()]`, `[equalTo()]`

More places below

File: packages/firebase_database/test/firebase_database_test.dart [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            'Query#keepSynced',
+            { 'path': 'foo', 'parameters': expectedParameters, 'value': true},
+          ),
+        ]));
+      });
+    });
+  });
+}
+
+class MockPlatformChannel extends Mock implements MethodChannel { }
```
Comment (mravn-google) [inline packages/firebase_database/test/firebase_database_test.dart]: Remove this?
And maybe end the file with a new-line.

File: packages/firebase_database/test/firebase_database_test.dart [hunk 11]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:mockito/mockito.dart';
```
Comment (mravn-google) [inline packages/firebase_database/test/firebase_database_test.dart]: Are you using `mockito` for anything? Seems your mock channel isn't used.

**Comment by collinjackson** [comment]
Thanks for the detailed feedback! I'll make these changes in a subsequent PR.

File: packages/firebase_database/lib/src/firebase_database.dart (hunk 5 above)
Comment (collinjackson) [inline packages/firebase_database/lib/src/firebase_database.dart]: If we pass primitive values here, we'll get a runtime type exception in Java when we say call.arguments(), at the top of the onMethodCall handler.

**Comment by collinjackson** [comment]
Remaining code review feedback implemented in https://github.com/flutter/plugins/pull/144

File: packages/firebase_database/lib/src/firebase_database.dart (hunk 5 above)
Comment (mravn-google) [inline packages/firebase_database/lib/src/firebase_database.dart]: You could let each branch of the switch handle arguments in its own way. Some of them don't use arguments at all, making the call to `arguments()`, the type check, and the assignment, superfluous for them.

Comment (collinjackson) [inline packages/firebase_database/lib/src/firebase_database.dart]: ok. I was optimizing for fewer lines of code but I see your point. I'll make the change.

**Pull request merged**